File: frontend/src/services/hyperv.ts
import { apiClient } from "../utils/apiClient";

const API = "/api/v1/hyperv";

export interface HyperVHost {
    id: number;
    name: string;
    host: string;
}

export interface HyperVVm {
    id: string;
    name: string;
    state: string;
    cpu_count: number;
    memory_assigned_mb: number;
    memory_startup_mb: number;
    memory_demand_mb: number;
    uptime_seconds: number;
    host_server: string;
    guest_os: string;
    creation_time: string | null;
    last_checkpoint: string | null;
    status_message: string;
    integration_services_enabled: boolean;
    cpu_usage_percent: number;
    disk_read_mbps: number;
    disk_write_mbps: number;
    network_receive_mbps: number;
    network_send_mbps: number;
}

export interface HyperVNetwork {
    id: string;
    name: string;
    switch_type: string;
    vlan_id: number | null;
    mac_address_spoofing: boolean;
    allow_management_os: boolean;
    status: string;
    connected_vms: number;
    net_adapter: string | null;
}

export interface HyperVStorage {
    id: string;
    name: string;
    path: string;
    size_bytes: number;
    used_bytes: number;
    type: string;
    vm_name: string | null;
    attached: boolean;
    format: string | null;
}

export interface HyperVCheckpoint {
    id: string;
    name: string;
    vm_name: string;
    vm_id: string | null;
    checkpoint_type: string;
    creation_time: string | null;
    size_bytes: number;
    parent_checkpoint_id: string | null;
    notes: string | null;
}

export interface HyperVSummary {
    connected: boolean;
    hostname: string | null;
    version: string | null;
    total_vms: number;
    running: number;
    stopped: number;
    paused: number;
    saved: number;
    total_cpu: number;
    total_memory_gb: number;
    used_memory_gb: number;
    total_storage_gb: number;
    used_storage_gb: number;
    host_servers: string[];
    network_switches: number;
    total_checkpoints: number;
    error: string | null;
}

export interface HyperVHealthHost {
    name: string;
    status: string;
    cpu_percent: number;
    memory_percent: number;
    memory_used_gb: number;
    memory_total_gb: number;
    uptime_seconds: number;
    vm_count: number;
    version: string | null;
}

export interface HyperVHealth {
    connected: boolean;
    status: string;
    hosts: HyperVHealthHost[];
    cluster_summary: string | null;
    error: string | null;
}

export interface HyperVActionResponse {
    success: boolean;
    message: string | null;
    error: string | null;
    vm_name: string | null;
}

export interface HyperVReplicationItem {
    vm_name: string;
    replica_server: string;
    replica_port: number;
    state: string;
    health: string;
    frequency_seconds: number;
    last_replication_time: string | null;
    last_result_code: number;
    bytes_sent: number;
    bytes_received: number;
}

export interface HyperVReplication {
    connected: boolean;
    replicating: number;
    total: number;
    items: HyperVReplicationItem[];
    error: string | null;
}

function hostParam(hostId: number | null): string {
    return hostId ? `?host_id=${hostId}` : "";
}

function hostParamAmp(hostId: number | null): string {
    return hostId ? `&host_id=${hostId}` : "";
}

export const hypervApi = {
    async listHosts(): Promise<HyperVHost[]> {
        const data = await apiClient<{ hosts: HyperVHost[] }>(`${API}/hosts`);
        return data.hosts ?? [];
    },

    async getSummary(hostId: number | null = null): Promise<HyperVSummary> {
        return apiClient<HyperVSummary>(`${API}/overview${hostParam(hostId)}`);
    },

    async getHealth(hostId: number | null = null): Promise<HyperVHealth> {
        return apiClient<HyperVHealth>(`${API}/health${hostParam(hostId)}`);
    },

    async testConnection(hostId: number | null = null): Promise<{ connected: boolean; latency_ms: number; message: string | null; hostname: string | null; error: string | null }> {
        return apiClient(`${API}/test${hostParam(hostId)}`);
    },

    async listVms(hostId: number | null = null): Promise<HyperVVm[]> {
        const data = await apiClient<{ items: HyperVVm[] }>(`${API}/vms${hostParam(hostId)}`);
        return data.items ?? [];
    },

    async getVm(vmId: string, hostId: number | null = null): Promise<HyperVVm> {
        const data = await apiClient<{ item: HyperVVm }>(`${API}/vms/${vmId}${hostParamAmp(hostId)}`);
        return data.item;
    },

    async startVm(vmId: string, hostId: number | null = null): Promise<HyperVActionResponse> {
        return apiClient<HyperVActionResponse>(`${API}/vms/${vmId}/start${hostParam(hostId)}`, { method: "POST" });
    },

    async stopVm(vmId: string, force = false, hostId: number | null = null): Promise<HyperVActionResponse> {
        return apiClient<HyperVActionResponse>(`${API}/vms/${vmId}/stop?force=${force}${hostParamAmp(hostId)}`, { method: "POST" });
    },

    async restartVm(vmId: string, hostId: number | null = null): Promise<HyperVActionResponse> {
        return apiClient<HyperVActionResponse>(`${API}/vms/${vmId}/restart${hostParam(hostId)}`, { method: "POST" });
    },

    async pauseVm(vmId: string, hostId: number | null = null): Promise<HyperVActionResponse> {
        return apiClient<HyperVActionResponse>(`${API}/vms/${vmId}/pause${hostParam(hostId)}`, { method: "POST" });
    },

    async resumeVm(vmId: string, hostId: number | null = null): Promise<HyperVActionResponse> {
        return apiClient<HyperVActionResponse>(`${API}/vms/${vmId}/resume${hostParam(hostId)}`, { method: "POST" });
    },

    async listNetworks(hostId: number | null = null): Promise<HyperVNetwork[]> {
        const data = await apiClient<{ items: HyperVNetwork[] }>(`${API}/networks${hostParam(hostId)}`);
        return data.items ?? [];
    },

    async listStorage(hostId: number | null = null): Promise<HyperVStorage[]> {
        const data = await apiClient<{ items: HyperVStorage[] }>(`${API}/storage${hostParam(hostId)}`);
        return data.items ?? [];
    },

    async listCheckpoints(vmId?: string, hostId: number | null = null): Promise<HyperVCheckpoint[]> {
        let url = `${API}/checkpoints`;
        const params: string[] = [];
        if (vmId) params.push(`vm_id=${vmId}`);
        if (hostId) params.push(`host_id=${hostId}`);
        if (params.length) url += `?${params.join("&")}`;
        const data = await apiClient<{ items: HyperVCheckpoint[] }>(url);
        return data.items ?? [];
    },

    async createCheckpoint(vmId: string, name?: string, hostId: number | null = null): Promise<{ success: boolean; message: string | null; error: string | null }> {
        let url = `${API}/checkpoints`;
        if (hostId) url += `?host_id=${hostId}`;
        return apiClient(url, {
            method: "POST",
            json: { vm_id: vmId, name },
        });
    },

    async deleteCheckpoint(vmId: string, checkpointId: string, hostId: number | null = null): Promise<{ success: boolean; message: string | null; error: string | null }> {
        return apiClient(`${API}/vms/${vmId}/checkpoints/${checkpointId}${hostParam(hostId)}`, { method: "DELETE" });
    },

    async getReplication(hostId: number | null = null): Promise<HyperVReplication> {
        return apiClient<HyperVReplication>(`${API}/replication${hostParam(hostId)}`);
    },
};

File: backend/app/schemas/hyperv.py
"""
Mission Control Hyper-V Schemas

Pydantic request/response models for the Hyper-V API.
"""

from pydantic import BaseModel

# ------------------------------------------------------------------ #
# VM                                                                  #
# ------------------------------------------------------------------ #


class HyperVVmResponse(BaseModel):
    id: str
    name: str
    state: str
    cpu_count: int
    memory_assigned_mb: int
    memory_startup_mb: int
    memory_demand_mb: int
    uptime_seconds: int
    host_server: str
    guest_os: str
    creation_time: str | None = None
    last_checkpoint: str | None = None
    status_message: str
    integration_services_enabled: bool
    cpu_usage_percent: float = 0.0
    disk_read_mbps: float = 0.0
    disk_write_mbps: float = 0.0
    network_receive_mbps: float = 0.0
    network_send_mbps: float = 0.0


class HyperVVmListResponse(BaseModel):
    count: int
    items: list[HyperVVmResponse]


class HyperVVmDetailResponse(BaseModel):
    item: HyperVVmResponse


class HyperVVmActionResponse(BaseModel):
    success: bool
    message: str | None = None
    error: str | None = None
    vm_name: str | None = None


# ------------------------------------------------------------------ #
# Network                                                             #
# ------------------------------------------------------------------ #


class HyperVNetworkResponse(BaseModel):
    id: str
    name: str
    switch_type: str
    vlan_id: int | None = None
    mac_address_spoofing: bool = False
    allow_management_os: bool = True
    status: str
    connected_vms: int = 0
    net_adapter: str | None = None


class HyperVNetworkListResponse(BaseModel):
    count: int
    items: list[HyperVNetworkResponse]


# ------------------------------------------------------------------ #
# Storage                                                             #
# ------------------------------------------------------------------ #


class HyperVStorageResponse(BaseModel):
    id: str
    name: str
    path: str
    size_bytes: int
    used_bytes: int
    type: str
    vm_name: str | None = None
    attached: bool
    format: str | None = None


class HyperVStorageListResponse(BaseModel):
    count: int
    items: list[HyperVStorageResponse]


# ------------------------------------------------------------------ #
# Checkpoint                                                          #
# ------------------------------------------------------------------ #


class HyperVCheckpointResponse(BaseModel):
    id: str
    name: str
    vm_name: str
    vm_id: str | None = None
    checkpoint_type: str
    creation_time: str | None = None
    size_bytes: int
    parent_checkpoint_id: str | None = None
    notes: str | None = None


class HyperVCheckpointListResponse(BaseModel):
    count: int
    items: list[HyperVCheckpointResponse]


class HyperVCheckpointCreateRequest(BaseModel):
    vm_id: str
    name: str | None = None


class HyperVCheckpointActionResponse(BaseModel):
    success: bool
    message: str | None = None
    error: str | None = None
    checkpoint: dict | None = None


# ------------------------------------------------------------------ #
# Hosts                                                               #
# ------------------------------------------------------------------ #


class HyperVHostInfo(BaseModel):
    id: int
    name: str
    host: str


class HyperVHostListResponse(BaseModel):
    hosts: list[HyperVHostInfo]


# ------------------------------------------------------------------ #
# Summary & Health                                                    #
# ------------------------------------------------------------------ #


class HyperVSummaryResponse(BaseModel):
    connected: bool
    hostname: str | None = None
    version: str | None = None
    total_vms: int = 0
    running: int = 0
    stopped: int = 0
    paused: int = 0
    saved: int = 0
    total_cpu: int = 0
    total_memory_gb: float = 0.0
    used_memory_gb: float = 0.0
    total_storage_gb: float = 0.0
    used_storage_gb: float = 0.0
    host_servers: list[str] = []
    network_switches: int = 0
    total_checkpoints: int = 0
    error: str | None = None


class HyperVHealthHostResponse(BaseModel):
    name: str
    status: str
    cpu_percent: float = 0.0
    memory_percent: float = 0.0
    uptime_seconds: int = 0
    vm_count: int = 0
    version: str | None = None


class HyperVHealthResponse(BaseModel):
    connected: bool
    status: str
    hosts: list[HyperVHealthHostResponse] = []
    cluster_summary: str | None = None
    error: str | None = None


class HyperVConnectionTestResponse(BaseModel):
    connected: bool
    latency_ms: int = 0
    message: str | None = None
    version: str | None = None
    hostname: str | None = None
    error: str | None = None


# ------------------------------------------------------------------ #
# Replication                                                         #
# ------------------------------------------------------------------ #


class HyperVReplicationItem(BaseModel):
    vm_name: str
    replica_server: str
    replica_port: int = 443
    state: str
    health: str
    frequency_seconds: int = 0
    last_replication_time: str | None = None
    last_result_code: int = 0
    bytes_sent: int = 0
    bytes_received: int = 0


class HyperVReplicationResponse(BaseModel):
    connected: bool
    replicating: int = 0
    total: int = 0
    items: list[HyperVReplicationItem] = []
    error: str | None = None
